tests/core/unittests/mongodb_test.py

```diff
@@ -0,0 +1,394 @@
+    yield database
+    client.close()
+
+
+@pytest.fixture()
+def clean_db(database, exp_config):
+    """Clean insert example experiment entries to collections."""
+    database.experiments.drop()
+    database.experiments.insert_many(exp_config[0])
+    database.trials.drop()
+    database.trials.insert_many(exp_config[1])
+    database.workers.drop()
+    database.workers.insert_many(exp_config[2])
+    database.resources.drop()
+    database.resources.insert_many(exp_config[3])
+
+
+@pytest.fixture(scope='module')
+def moptdb():
+    """Return MongoDB wrapper instance initiated with test opts."""
+    moptdb = MongoDB(username='user', password='pass', dbname='metaopt_test')
+    return moptdb
+
+
+class TestConnection(object):
+    """Create a :class:`metaopt.io.database.MongoDB`, check connection cases."""
+
+    def test_bad_connection(self):
+        """Raise when connection cannot be achieved."""
+        with pytest.raises(DatabaseError) as exc_info:
+            MongoDB(host='asdfada', username='uasdfaf', password='paasdfss')
+        assert "Connection" in str(exc_info.value)
+
+    def test_bad_authentication(self):
+        """Raise when authentication cannot be achieved."""
+        with pytest.raises(DatabaseError) as exc_info:
+            MongoDB(dbname='metaopt_test', username='uasdfaf', password='paasdfss')
+        assert "Authentication" in str(exc_info.value)
+
+    def test_connection_with_uri(self):
+        """Check the case when connecting with ready `uri`."""
+        moptdb = MongoDB(uri='mongodb://user:pass@localhost/metaopt_test')
+        assert moptdb.host == 'localhost'
+        assert moptdb.port == 27017
+        assert moptdb.username == 'user'
+        assert moptdb.password == 'pass'
+        assert moptdb.dbname == 'metaopt_test'
+
+
+@pytest.mark.usefixtures("clean_db")
+class TestRead(object):
+    """Calls to :meth:`metaopt.io.database.MongoDB.read`."""
+
+    def test_read_experiment(self, exp_config, moptdb):
+        """Fetch a whole experiment entries."""
+        loaded_config = moptdb.read(
+            'experiments', {'exp_name': 'supernaedo2', 'metadata.user': 'tsirif'})
+        assert loaded_config == exp_config[0][:2]
+
+        loaded_config = moptdb.read(
+            'experiments', {'exp_name': 'supernaedo2', 'metadata.user': 'tsirif'},
+            atomic=True)
+        assert loaded_config == exp_config[0][1]
+
+        loaded_config = moptdb.read('experiments',
+                                    {'exp_name': 'supernaedo2',
+                                     'metadata.user': 'tsirif',
+                                     'metadata.datetime': datetime(2017, 11, 22, 23, 0, 0)})
+        assert loaded_config == exp_config[0][0]
+
+    @pytest.mark.skip(reason="sugar")
+    def test_ctx_read_experiment(self, exp_config, moptdb):
+        """Fetch a whole experiment entry using an experiment context (always atomic)."""
+        with moptdb('experiments', {'exp_name': 'supernaedo2',
```

Comment:
Yes I saw the sugar skip. :laughing: 

If I may ask you, could you please remove tests that are for sugar features. You can keep them in a branch on your fork to avoid loosing the code. I would also encourage you to make an issue where you explain your idea of context manager if you do spend some time on that. Nevertheless, we are past the milestone for the DB and are gonna soon be past the milestone for Experiment/Trials so I would recommend we focus on the base first.